### PR TITLE
Add a new boolean flag to IIsTypeExpression to represent VB TypeOf Is…

### LINKED ISSUE #8882: IOperation: VB - It looks like there is no way to represent [TypeOf ... IsNot ...] via IIsExpression
It looks like, given the current implementation, [TypeOf ... **IsNot** ...] and [TypeOf ... **Is** ...] are represented the same way and the consumer is unable to distinguish them.

### PR DESCRIPTION
**Pull request by mavasani**
…Not expression

Fixes #8882

### CODE REVIEW

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIsTypeExpression.cs [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+namespace Microsoft.CodeAnalysis.CSharp.UnitTests
+{
+    public partial class IOperationTests : SemanticModelTestBase
+    {
+        [CompilerTrait(CompilerFeature.IOperation)]
+        [Fact]
+        public void TestIsOperator()
+        {
+            string source = @"
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIsTypeExpression.cs]: Could you break up these tests? Large test blocks bring in a lot of unrelated IOperations and it is both hard to review, and can introduce a lot of meaningless noise when unrelated nodes change shape.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIsTypeExpression.cs]: > Could you break up these tests?

I support this request as well. Please use more targeted tests.

**Comment by AlekseyTs** [comment]
> namespace Microsoft.CodeAnalysis.CSharp.UnitTests

It looks like there are no meaningful changes in this file, please revert it to original state.

---
Refers to: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IWhileUntilLoopStatement.cs:9 in f0b2b35. [](commit_id = f0b2b35681f83776673cbde81b3189fdf11084fd, deletion_comment = False)

File: src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs [hunk 2]
```diff
@@ -880,12 +880,18 @@ public override void VisitCoalesceExpression(ICoalesceExpression operation)
...
+            {
+                LogString(" (IsNotExpression)");
+            }
+
             LogCommonPropertiesAndNewLine(operation);
 
             Visit(operation.Operand, "Operand");
 
             Indent();
-            LogType(operation.Type);
```
Comment (AlekseyTs) [inline src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs]: >LogType(operation.Type); [](start = 12, length = 24)

It feels like we should keep Type property in the output.

File: src/Compilers/VisualBasic/Test/Semantic/Semantics/TypeOfTests.vb [hunk 3]
```diff
@@ -211,12 +409,287 @@ Module Program
...
 End Module
-    </file>
-</compilation>
-
-            Dim compilation = CompilationUtils.CreateCompilationWithMscorlibAndVBRuntimeAndReferences(source, , TestOptions.ReleaseExe)
-            CompilationUtils.AssertTheseDiagnostics(compilation,
-<expected>
+]]>.Value
+
+            Dim expectedOperationTree = <![CDATA[
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Semantic/Semantics/TypeOfTests.vb]: >expectedOperationTree [](start = 16, length = 21)

Please adjust tests to target specific expressions instead. Also, consider limiting testing to distinct scenarios that actually interesting from IOperation point of view.

**Comment by AlekseyTs** [comment]
Done with review pass.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIsTypeExpression.cs (hunk 1 above)
Comment (mavasani) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIsTypeExpression.cs]: Test has been ported from http://source.roslyn.io/#Roslyn.Compilers.CSharp.Emit.UnitTests/CodeGen/CodeGenOperators.cs,63ad66a1ca6ab689. Ideally, I would have just enhanced the existing test to have the operation tree verification, but the test lies in compiler emit tests project, and I think it is best to keep the operation tests in semantic tests project. Is it fine to clone and break down these tests into individual operation tests?

File: src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs (hunk 2 above)
Comment (mavasani) [reply src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs]: `LogCommonPropertiesAndNewLine` already logs the type.

File: src/Compilers/VisualBasic/Test/Semantic/Semantics/TypeOfTests.vb (hunk 3 above)
Comment (mavasani) [reply src/Compilers/VisualBasic/Test/Semantic/Semantics/TypeOfTests.vb]: I am slightly confused - I thought we were specifically asked to extend existing compiler semantic tests to avoid duplicating test code and verification wherever possible. Is our preference now always to create new tests for Operation tree verification?

Comment (AlekseyTs) [reply src/Compilers/VisualBasic/Test/Semantic/Semantics/TypeOfTests.vb]: > Is our preference now always to create new tests for Operation tree verification

I didn't suggest to create new tests. I suggested to make the tests more targeted and readable for humans. I.e. don't dump IOperation for arbitrary big chunks of code and expect reviewers to verify them. Why do you include entire declaration statements, ```if``` statements and the like into the base line? If I were you, in this PR I would verify IOperation only for the TypeOf expressions.

Comment (mavasani) [reply src/Compilers/VisualBasic/Test/Semantic/Semantics/TypeOfTests.vb]: @333fred Will your tool be able to handle updating tests written in this manner? Otherwise, we might have to hand fix all these tests everytime the tree shape changes.

Comment (333fred) [reply src/Compilers/VisualBasic/Test/Semantic/Semantics/TypeOfTests.vb]: If you mean putting multiple asserts in a single test function, it won't be able to handle that. We could make it do something like that, but it would probably be more complex than useful.

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIsTypeExpression.cs (hunk 1 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIsTypeExpression.cs]: I would expect to see tests that only verify bits of IOperation tree that are relevant to the change. This can be achieved in many different ways. For example, requesting and verifying IOperation for several specific syntax nodes in context of the same test is acceptable. What is not acceptable in my opinion, is to dump tree for big chunks of code that include all kind of constructs unrelated to the change. The main purpose of the added tests should be to verify effect of the change, the used approach makes it nearly impossible as the baseline is extremely hard to comprehend.

Comment (mavasani) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIsTypeExpression.cs]: @AlekseyTs I completely understand your request and the desire to have targeted tests. However, such tests are extremely hard to update for baseline changes, which are happening extremely frequently in our current development cycle. @333fred has a written a tool that bulk edits the baselines for tests written in a standard format - single operation tree verification in a unit test. If we start to hand craft tests where it is verifying multiple small pieces of operation tree, then we will need to spend lot of manual effort just updating the baselines for these tests, which is currently happening for almost every other merge conflict resolution. I would like to propose the following approach:

1. Search for existing compiler binding tests in the feature area
   1. If the tests exist and are small and targeted tests, extend them to verify the operation tree.
   2. If the tests exist but are large tests (like the TypeOf tests in this PR), we clone and break them into separate targeted IOperation tests.
   3. If the tests don't exist or are scattered all over the test projects, we create new IOperation unit tests.

So for the current PR, I would like to take approach 2.

**Comment by mavasani** [comment]
@AlekseyTs @333fred I have broken down the tests into smaller targeted tests.

**Comment by AlekseyTs** [comment]
> namespace Microsoft.CodeAnalysis.CSharp.UnitTests

It looks like there are no meaningful changes in this file, please revert it to the original state and remove it from this PR.

---
Refers to: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IWhileUntilLoopStatement.cs:9 in b10cac4. [](commit_id = b10cac4ea34b28bd991d38b1e2d22210361a69c7, deletion_comment = False)

**Review by AlekseyTs — APPROVED** [review]
LGTM

**Comment by mavasani** [comment]
> It looks like there are no meaningful changes in this file, please revert it to the original state and remove it from this PR.

I am not sure why git is showing changes to this file. I tried copying the entire file header from features/ioperation into my local file on this branch and it shows no diff. Anyways, this file is changing quite a bit in my other PR for loops.
